kame/kame/hroute6d/admin/rip6admin.8
.\" Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     $Id: rip6admin.8,v 1.1 1999/08/08 23:29:37 itojun Exp $
.\"
.Dd May 19, 1998
.Dt RIP6ADMIN 8
.Os KAME
.\"
.Sh NAME
.Nm rip6admin
.Nd administrative utility for network routing daemon, hroute6d
.\"
.Sh SYNOPSIS
.Nm rip6admin
.Ar command
.\"
.Sh DESCRIPTION
.Nm Rip6admin
is invoked by system administrator to control the
.Xr hroute6d 8 .
.Pp
The
.Nm rip6admin
utility provides 7 commands:
.Pp
.Bl -tag -width Fl -compact
.It Cm restart
Make hroute6d to read configuration file
.Xr hroute6d.conf 5
and start again.
.It Cm stop
Clear kernel routing table and kill the hroute6d daemon.
.It Cm trace
The hroute6d daemon can trace the incoming and outgoing packets based on
the value of a flag. The flag can be toggled between on/off using this
command.
.It Cm reset
Hroute6d maintains statistics for number of bad packets, queries received etc. 
These counters can be reset to zero using this command.
.It Cm stat
Display current statistics counters.
.It Cm table Ar prefix prefixlen 
Display hroute6d's local routing information.
.It Cm start
Start hroute6d daemon. Do not use when hroute6d is running.
.El
.Pp
The table command has the syntax
.Pp
.Bd -filled -offset indent -compact
.Nm rip6admin
.Cm table
.Ar prefix
.Ar prefixlen
.Ed
.Pp
Routing Informations that match
.Ar prefix
.Ar prefixlen
are displayed. If no routes matches, nothing is displayed.
To see 'default' route, type
.Pp
.Bd -filled -offset indent -compact
.Nm rip6admin
.Cm table
.Ar default .
.Ed
.Pp
To see all informations, type
.Pp
.Bd -filled -offset indent -compact
.Nm rip6admin
.Cm table
.Ar ::
.Ar 0
.Ed
.Pp
(0 prefixlen matches all route informations).
.Pp
The counters maintained by hroute6d are 
.Bl -bullet -compact
.It
global route changes counter
.It
received valid queries counter
.It
received bad packets counter
.It
received bad RTEs counter
.It
sent updates counter
.El
.\"
.Sh SEE ALSO
.Xr rip6query 8 ,
.Xr hroute6d 8


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/hroute6d/admin/rip6admin.c
/* 
 * $Id: rip6admin.c,v 1.1.1.1 1999/08/08 23:29:37 itojun Exp $
 */

/*
 * Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * Copyright(C)1997 by Hitach, Ltd.
 */

#include "rip6adm.h"
#include "pathnames.h"

/* 
 * Takes the command line parameters and communicates appropriately
 * with the route6 daemon.
 */
int
main(int argc, char *argv[])
{
	int command;
	int pid;
	struct sigaction sig;

	bzero((void *)&sig, sizeof(sig));
	sig.sa_handler = (void *)cleanup;
	sig.sa_flags = 0;
	if (sigaction(SIGALRM, &sig, (struct sigaction *)NULL) < 0) {
		perror("sigaction");
		exit(1);
	}
	sig.sa_flags = 0;
	sig.sa_handler = (void *)cleanup;
	if (sigaction(SIGTERM, &sig, (struct sigaction *)NULL) < 0) {
		perror("sigaction");
		exit(1);
	}
	if ((command = parse(argc, argv)) < 0) {
		usage();
		exit(1);
	}
	switch (command) {
	case ADM_STAT:
	case ADM_TABLE:
		process_req();
		break;

	case ADM_SIGNAL:
		/* superuser permissions reqd. to send signals to route6d */
		if (getuid() != 0) {
			fprintf(stderr, "Permission denied\n");
			exit(1);
		}
		if ((pid = getroute6dpid()) == 0) {
			fprintf(stderr, "%s is not runnig.", PROGNAME);
			exit(1);
		}

		/* Send requested signal to route6d */
		if (kill(pid, sigval) < 0)
			perror("kill");
		break;

	case ADM_EXEC:
		if (getuid() != 0) {
			fprintf(stderr, "Permission denied\n");
			exit(1);
		}
		if ((pid = getroute6dpid()) != 0) {
			fprintf(stderr, "maybe %s is still alive\n", PROGNAME);
			exit(1);
		}
		execl(RT6_PATH, RT6_PATH, NULL);
		perror("start failed");
		break;
	}

	cleanup();
	exit(0);
}

/* 
 * Parse the command line to check for illegal usage or invalid
 * parameters. If the command is one of [restart | reset |
 * trace | stop] then identify the respective signal to be sent. For
 * stat or table the request is constructed.
 */
int
parse(int argc, char *argv[])
{
	int len = 0;
	int i;

	bzero((void *)&infodetail, sizeof(struct info_detail));
	if (argc < 2)
		return(-1);

	if (argc > 2) {
		if (((strcasecmp("table", argv[1])) != 0) || (argc > 4))
			return(-1);

		if (argc == 3 && (strcasecmp("default", argv[2]) != 0))
			return(-1);

		infodetail.id_type = ADM_TABLE;

		if (argc == 3) {
			len = infodetail.id_prflen = 128;
			bzero((void *)infodetail.id_addr.s6_addr,
			      sizeof(struct in6_addr));
			return (infodetail.id_type);
		} else {
			len = atoi(argv[3]);
			if (!(VALID_PREF_ADDR_LEN(len)))
				return (-1);
			infodetail.id_prflen = (u_char) len;
		}

		/* if request for entire routing table len is 0 */
		if (len == 0)
			bzero(&infodetail.id_addr, sizeof(struct in6_addr));
		else
			if ((inet_pton(AF_INET6, argv[2],
				       (void *)&(infodetail.id_addr))) <= 0)
				return(-1);

		i = (len - 1) / 8 + 1;
		infodetail.id_addr.s6_addr[i - 1] &=
			((signed char)0x80 >> ((len - 1) % 8));
		for (; i < 16; i++)
			infodetail.id_addr.s6_addr[i] = 0;

		return(infodetail.id_type);
	}
	if ((strcasecmp("stat", argv[1])) == 0) {
		infodetail.id_type = ADM_STAT;
		return(infodetail.id_type);
	}
	if ((strcasecmp("restart", argv[1])) == 0) {
		sigval = SIGHUP;
		return(ADM_SIGNAL);
	}
	if ((strcasecmp("reset", argv[1])) == 0) {
		sigval = SIGINT;
		return(ADM_SIGNAL);
	}
	if ((strcasecmp("trace", argv[1])) == 0) {
		sigval = SIGUSR1;
		return(ADM_SIGNAL);
	}
	if ((strcasecmp("stop", argv[1])) == 0) {
		sigval = SIGTERM;
		return(ADM_SIGNAL);
	}
	if ((strcasecmp("start", argv[1])) == 0) {
		return(ADM_EXEC);
	}
	return (-1);
}

/* 
 * Display the correct usage
 */
void
usage(void)
{
	fprintf(stderr, "Usage : rip6admin [restart | reset | trace | stop]\n");
	fprintf(stderr, "\t\t  [stat] [table prefix prefixlen]\n\n");
	return;
}

/* 
 * Gets the process id of route6d from pid file
 */
int
getroute6dpid(void)
{
	FILE *fp;
	int pid = 0;

	if ((fp = fopen(RT6_PID, "r")) == NULL)
		return(0);
	fscanf(fp, "%d", &pid);
	fclose(fp);
	return(pid);
}

/* 
 * Send the request for statistics or routing table to route6d over
 * UNIX domain sockets.
 */
void
process_req(void)
{
	int sockfd, firstpkt = 1, numofentries, datalen, len, optval;
	u_char databuf[ADM_PKTSIZE];
	char prefstr[INET6_ADDRSTRLEN];
	char *offset, line[100], line2[100];
	struct sockaddr_un localsock, route6dsock;
	struct sockaddr_un src;
	struct statistic *stat;
	struct per_if_info *ifinfo;
	struct rt_table *rte;

	if ((sockfd = socket(AF_UNIX, SOCK_DGRAM, 0)) < 0) {
		perror("socket");
		return;
	}
	optval = ADM_BUFSIZE;
	if (setsockopt(sockfd, SOL_SOCKET, SO_RCVBUF,
		       (void *)&optval, sizeof(int))) {
		perror("setsockopt");
		return;
	}
	localsock.sun_family = AF_UNIX;

	/* 
	 * To allow multiple rip6admin commands to run, bind to a dynamically 
	 * constructed path.
	 */
	bzero((void *)udspath, PATH_MAX);
	sprintf(udspath, "%s%d", ADM_RIP6_ADM, getpid());
	strcpy(localsock.sun_path, udspath);
	if (bind(sockfd, (struct sockaddr *)&localsock,
		 sizeof(struct sockaddr_un))) {
		perror("bind");
		return;
	}
	route6dsock.sun_family = AF_UNIX;
	strcpy(route6dsock.sun_path, ADM_RIP6_UDS);

	if ((len = sendto(sockfd, &infodetail, sizeof(infodetail), 0,
	   (struct sockaddr *)&route6dsock, sizeof(struct sockaddr_un))) < 0
	    || len != (sizeof(infodetail))) {
		perror("sendto");
		return;
	}
	while (1) {
		len = sizeof(struct sockaddr_un);
		alarm(TIMEOUT);
		if ((datalen = recvfrom(sockfd, databuf, ADM_PKTSIZE, 0,
				      (struct sockaddr *)&src, &len)) < 0) {
			perror("recvfrom");
			return;
		}
		/* deactivate the alarm */
		alarm(0);
		/*
		 * End of data = packet of len 1 with ADM_EOF as the
                 * only char in it
		 */
		if (datalen == 1 && databuf[0] == ADM_EOF)
			break;

		numofentries = 0;
		switch (infodetail.id_type) {
		case ADM_STAT:
			/* 
			 * reply pkt. contain the global counts first
			 followed by per interface stats.
			 */
			if (firstpkt) {
				numofentries = (datalen - sizeof(struct statistic)) /
					sizeof(struct per_if_info);
				firstpkt = 0;
				if (datalen < sizeof(struct statistic))
					 return;
				stat = (struct statistic *)databuf;
				printf("Routes changed:\t%lu\n", stat->st_grccount);
				printf("Valid queries:\t%lu\n", stat->st_gqcount);
				offset = 0;
				offset = (char *)databuf + sizeof(struct statistic);
				ifinfo = (struct per_if_info *)offset;
				printf("Interface\t\tBad Pkts\t\tBad RTEs\t\tSndPkts\n");
			} else {
				numofentries = datalen / sizeof(struct per_if_info);
				ifinfo = (struct per_if_info *)databuf;
			}
			while (numofentries--) {
				printf("  %s\t\t\t", ifinfo->pi_ifname);
				printf("%lu\t\t\t", ifinfo->pi_badpkt);
				printf("%lu\t\t\t", ifinfo->pi_badrte);
				printf("%lu\n", ifinfo->pi_updates);
				ifinfo++;
			}
			break;

		case ADM_TABLE:
			numofentries = datalen / sizeof(struct rt_table);
			rte = (struct rt_table *)databuf;
			while (numofentries--) {
				if (inet_ntop(AF_INET6, (void *)&(rte->rt_dest),
					      prefstr, sizeof(prefstr)) == 0) {
					strcpy(line, "<<Incorrect data>>");
				} else
					sprintf(line, "%s/%d", prefstr,
						rte->rt_prflen);

				
				if (inet_ntop(AF_INET6, (void *)&(rte->rt_gway),
					      prefstr, sizeof(prefstr)) == 0) {
					strcpy(line2, "<<Incorrect data>>");
				} else
					strcpy(line2, prefstr);

				printf("%-43s %s\n", line, line2);
				printf("\t%-5s metric %d ",
				       rte->rt_ifname, rte->rt_metric);
				displayflg(rte->rt_flag);
				
				rte++;
			}
			break;
		}
	}
	close(sockfd);
	unlink(udspath);
	return;
}

/* 
 * Displays the flags corresponding to bits set.
 */
void
displayflg(int flag)
{
	int f = 0;
	int tryflag = 1;

	printf("< ");
	while (flag) {
		f = flag & tryflag;
		flag &= ~tryflag;
		tryflag <<= 1;
		switch (f) {
		case RTS6_CHANGED:
			printf("CHANGED ");
			break;
		case RTS6_STATIC:
			printf("STATIC ");
			break;
		case RTS6_DEFAULT:
			printf("DEFAULT ");
			break;
		case RTS6_KERNEL:
			printf("KERNEL ");
			break;
		case RTS6_LOOPBACK:
			printf("LOOPBACK ");
			break;
		case RTS6_INTERFACE:
			printf("INTERFACE ");
			break;
		case RTS6_PTOP:
			printf("PTOP ");
			break;
		}
	}
	printf(">\n");
	return;
}

void
cleanup(void)
{
	unlink(udspath);
	exit(1);
}
